backend/de.metas.printing/de.metas.printing.base/src/main/java/de/metas/printing/async/spi/impl/PrintingQueuePDFConcatenateWorkpackageProcessor.java

```diff
@@ -0,0 +1,117 @@
+import org.compiere.model.I_AD_Archive;
+import org.slf4j.Logger;
+import org.slf4j.MDC;
+
+import java.io.File;
+import java.io.FileOutputStream;
+import java.io.IOException;
+import java.util.List;
+
+/**
+ * This processor process the enqueued <code>I_C_Printing_Queue</code>
+ *
+ * <ul> for each C_Printing_Queue selection</ul>
+ * <ul> fetches the PDF files</ul>
+ * <ul> and concatenates all pdf in one big pdf file per selection</ul>
+ *
+ * @author cg
+ */
+public class PrintingQueuePDFConcatenateWorkpackageProcessor implements IWorkpackageProcessor
+{
+	private final Logger logger = LogManager.getLogger(PrintingQueuePDFConcatenateWorkpackageProcessor.class);
+
+	private final AttachmentEntryService attachmentEntryService = SpringContextHolder.instance.getBean(AttachmentEntryService.class);
+	private final IPrintingDAO dao = Services.get(IPrintingDAO.class);
+	private final IQueueDAO queueDAO = Services.get(IQueueDAO.class);
+	private final IPrintingQueueBL printingQueueBL = Services.get(IPrintingQueueBL.class);
+	private final IArchiveBL archiveBl = Services.get(IArchiveBL.class);
+
+	private I_C_Async_Batch asyncBatch;
+
+	@Override
+	public Result processWorkPackage(final I_C_Queue_WorkPackage workpackage, final String localTrxName_NOTUSED)
+	{
+		File outputFile = null;
+		try
+		{
+			outputFile = concatenateFiles(workpackage);
+		}
+		catch (IOException | DocumentException e)
+		{
+			Loggables.withLogger(logger, Level.ERROR).addLog(e.getLocalizedMessage());
+		}
+
+		attachmentEntryService.createNewAttachment(workpackage.getC_Async_Batch(), outputFile);
+		return Result.SUCCESS;
+	}
+
+	private File concatenateFiles(final I_C_Queue_WorkPackage workpackage) throws IOException, DocumentException
+	{
+		this.asyncBatch = workpackage.getC_Async_Batch();
+		Check.assumeNotNull(asyncBatch, "Async batch is not null");
+
+		final String fileName = "PDF_" + asyncBatch.getC_Async_Batch_ID();
+
+		final File file = File.createTempFile(fileName, ".pdf");
+		final Document document = new Document();
+
+		final FileOutputStream fos = new FileOutputStream(file, false);
+		final PdfCopy copy = new PdfCopy(document, fos);
+
+		document.open();
+
+		final List<I_C_Printing_Queue> pqs = queueDAO.retrieveAllItems(workpackage, I_C_Printing_Queue.class);
+
+		for (final I_C_Printing_Queue pq : pqs)
+		{
+			try (final MDC.MDCCloseable ignored = TableRecordMDC.putTableRecordReference(pq))
+			{
+				if (pq.isProcessed())
+				{
+					Loggables.withLogger(logger, Level.DEBUG).addLog("*** Printing queue is already processed. Skipping it: {}", pq);
+					continue;
+				}
+				final I_AD_Archive archive = pq.getAD_Archive();
+				Check.assume(archive != null, pq + " references an AD_Archive record");
+
+				final byte[] data = archiveBl.getBinaryData(archive);
+				final PdfReader reader = new PdfReader(data);
+
+				for (int page = 0; page < reader.getNumberOfPages(); )
+				{
+					copy.addPage(copy.getImportedPage(reader, ++page));
```

Comment:
this look a bit weird.... mainly the `++page`.
Is this copied from somewhere? 
Why not consider refactoring the code and have a shared static method which appends a PdfReader to a PdfCopy?